python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java

```diff
@@ -49,6 +49,12 @@
 public class SymbolsModuleTypeProvider {
   private final ProjectLevelSymbolTable projectLevelSymbolTable;
   private final TypeShed typeShed;
+  private final Map<String, Map<String, PythonType>> syntheticModuleMembers = Map.ofEntries(
```

Comment:
I understand we need to do something given that `TypeShed` doesn't clearly define some `typing` protocols.
I'm not sure how we should deal with that, though. We actually have several options:
- Process known protocols directly in the `typeshed_serializer`, and serialize them in a more expected format
- Have a list such as this one that could be used when converting symbols to types without storing them in a separate`syntheticModuleMembers`; potentially, we could alter the `TYPE_SOURCE` as a result.
- Maintain 2 hierarchies such as these ones

I don't really have a preference, and I think the current approach is fine for now; I just think it's good to highlight these options.

That being said, I have some comments:
- If I remove this, unit tests (outside of those specifically testing this synthetic look-up feature) still seem to work. In particular, unit tests for `confusingTypeChecking/nonCallableCalled.py` still pass. I would suspect this is needed to avoid FPs, but then I would expect a reproducer for those FPs to be tested in the rule tests.

- Following on the first point, I would expect that missing symbol information (especially here where `typing.Callable` is defined in Typeshed as a simple variable) would not result in FPs (they should at most lead to FNs). Therefore, why do we need this?

- If we do need these synthetic types, why not give them the expected `__call__` member, so that we don't have to special-case them in the relevant check later?

EDIT: As I tested this further, it seems that the behavior when removing these synthetic types is actually nondeterministic. I suspect that what is inferred as `typing.Coroutine` and what is in the predicate may be different objects, therefore failing the check. It's true that the `typing.Coroutine`we retrieve lacks the `__call__` method either way, which is annoying and might warrant the special case. However, I'm not sure synthetic types are an appropriate fix for that.